### PR TITLE
adding AddingObserver event and RemovingObserver event for Scp173

### PR DESCRIPTION
**Pull request by skyyt15**
the adding observer is triggered when a user just saw scp 173. for example, if ev.IsAllowed is set to false and the user is the only one who can saw scp173, the peanut can keep moving (it will be not added in the Observers list)

an example when the ev.IsAllowed is set to false
image

he RemovingObserver is the same as the AddingObserver event but there is no a IsAllowed parameter (it's weird in this situation) and it's trigger when the player no longer saw scp173.

special thanks to Bolton who help me to find stupid IL errors
special thing to yamato: i have compile this xmas-2023 version in release build :trollface:

edit: ho i forgot to tell: i have build all change and tested the events with a different plugin in my personnal server

### CODE REVIEW

File: Exiled.Events/EventArgs/Scp173/AddingObserverEventArgs.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+namespace Exiled.Events.EventArgs.Scp173
+{
+    using API.Features;
+    using Interfaces;
+
+    /// <summary>
+    /// Contains all information before a players just saw scp 173 (useful for an UTR for example).
+    /// </summary>
+    public class AddingObserverEventArgs : IExiledEvent, IDeniableEvent
```
Comment (BoltonDev) [inline Exiled.Events/EventArgs/Scp173/AddingObserverEventArgs.cs:16]:
```suggestion
    public class AddingObserverEventArgs : IScp173Event, IDeniableEvent
```

File: Exiled.Events/EventArgs/Scp173/AddingObserverEventArgs.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+// </copyright>
+// -----------------------------------------------------------------------
+
+namespace Exiled.Events.EventArgs.Scp173
+{
+    using API.Features;
+    using Interfaces;
+
+    /// <summary>
+    /// Contains all information before a players just saw scp 173 (useful for an UTR for example).
```
Comment (BoltonDev) [inline Exiled.Events/EventArgs/Scp173/AddingObserverEventArgs.cs:14]:
```suggestion
    /// Contains all information before a player sees scp 173.
```
Remove example and doc rework

File: Exiled.Events/EventArgs/Scp173/AddingObserverEventArgs.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        {
+            Scp173 = scp173;
+            Target = target;
+            IsAllowed = isAllowed;
+        }
+
+        /// <summary>
+        /// Gets the scp 173 as a <see cref="Player"/>.
+        /// </summary>
+        public Player Scp173 { get; }
```
Comment (BoltonDev) [inline Exiled.Events/EventArgs/Scp173/AddingObserverEventArgs.cs:40]:
```suggestion
        /// <summary>
        /// Gets the player who controlling SCP-173.
        /// </summary>
        public Player Player { get; }

        /// <inheritdoc/>
        public Scp173Role Scp173 { get; }
```

File: Exiled.Events/EventArgs/Scp173/AddingObserverEventArgs.cs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        /// </param>
+        /// <param name="isAllowed">
+        /// <inheritdoc cref="IsAllowed" />
+        /// </param>
+        public AddingObserverEventArgs(Player scp173, Player target, bool isAllowed = true)
+        {
+            Scp173 = scp173;
+            Target = target;
+            IsAllowed = isAllowed;
+        }
```
Comment (BoltonDev) [inline Exiled.Events/EventArgs/Scp173/AddingObserverEventArgs.cs:35]:
```suggestion
        public AddingObserverEventArgs(Player player, Player target, bool isAllowed = true)
        {
            Player = player;
            Scp173 = player.Role.As<Scp173Role>();
            Target = target;
            IsAllowed = isAllowed;
        }
```

File: Exiled.Events/EventArgs/Scp173/AddingObserverEventArgs.cs [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        /// </summary>
+        public Player Scp173 { get; }
+
+        /// <summary>
+        /// Gets the target who saw the scp 173, as a <see cref="Player"/>.
+        /// </summary>
+        public Player Target { get; }
+
+        /// <summary>
+        /// Gets or Sets a value indicating whether or not the player trigger scp 173 (for exemple, if scp173 is alone with the Target, it's can keep moving).
```
Comment (BoltonDev) [inline Exiled.Events/EventArgs/Scp173/AddingObserverEventArgs.cs:48]:
```suggestion
        /// Gets or sets a value indicating whether or not the player triggers scp 173.
```
Remove example

File: Exiled.Events/EventArgs/Scp173/AddingObserverEventArgs.cs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+// -----------------------------------------------------------------------
+
+namespace Exiled.Events.EventArgs.Scp173
+{
+    using API.Features;
+    using Exiled.API.Features.Roles;
+    using Interfaces;
+
+    /// <summary>
+    /// Contains all information before a player sees scp 173.
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/Scp173/AddingObserverEventArgs.cs:15]:
```suggestion
    /// Contains all information before a player sees SCP-173.
```

File: Exiled.Events/EventArgs/Scp173/AddingObserverEventArgs.cs [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    /// </summary>
+    public class AddingObserverEventArgs : IScp173Event, IDeniableEvent
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AddingObserverEventArgs"/> class.
+        /// </summary>
+        /// <param name="player">
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/Scp173/AddingObserverEventArgs.cs:22]: Keep all `<param>`'s on the same line

File: Exiled.Events/EventArgs/Scp173/AddingObserverEventArgs.cs [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+            Player = player;
+            Target = target;
+            IsAllowed = isAllowed;
+        }
+
+        /// <inheritdoc />
+        public Scp173Role Scp173 { get; }
+
+        /// <summary>
+        /// Gets the target who saw the scp 173.
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/Scp173/AddingObserverEventArgs.cs:43]:
```suggestion
        /// Gets the target who have just seen SCP-173.
```

File: Exiled.Events/EventArgs/Scp173/AddingObserverEventArgs.cs [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+        /// </summary>
+        public Player Target { get; }
+
+        /// <summary>
+        /// Gets the player who controlling scp 173.
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/Scp173/AddingObserverEventArgs.cs:48]:
```suggestion
        /// Gets the player who's controlling SCP-173.
```

File: Exiled.Events/EventArgs/Scp173/AddingObserverEventArgs.cs [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        /// </summary>
+        public Player Player { get; }
+
+        /// <summary>
+        /// Gets or Sets a value indicating whether or not the player trigger scp 173.
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/Scp173/AddingObserverEventArgs.cs:53]:
```suggestion
        /// Gets or sets a value indicating whether the player can be added as an observer.
```

File: Exiled.Events/EventArgs/Scp173/RemovingObserverEventArgs.cs [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+// -----------------------------------------------------------------------
+
+namespace Exiled.Events.EventArgs.Scp173
+{
+    using API.Features;
+    using Exiled.API.Features.Roles;
+    using Interfaces;
+
+    /// <summary>
+    /// Contains all information before a player no longer see scp 173.
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/Scp173/RemovingObserverEventArgs.cs:15]:
```suggestion
    /// Contains all information after a player stops looking at SCP-173.
```

File: Exiled.Events/Patches/Events/Scp173/AddingObserver.cs [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+            LocalBuilder ev = generator.DeclareLocal(typeof(AddingObserverEventArgs));
+
+            int index = newInstructions.FindIndex(x => x.Calls(Method(typeof(HashSet<ReferenceHub>), nameof(HashSet<ReferenceHub>.Add)))) + 2;
+            newInstructions[index].WithLabels(returnLabel);
+            newInstructions.InsertRange(index, new CodeInstruction[]
+            {
+                // Player.Get(Owner);
+                new(OpCodes.Ldarg_0),
+                new(OpCodes.Callvirt,
+                    PropertyGetter(typeof(StandardSubroutine<Scp173Role>), nameof(StandardSubroutine<Scp173Role>.Owner))),
```
Comment (NaoUnderscore) [inline Exiled.Events/Patches/Events/Scp173/AddingObserver.cs:45]: Keep it on the same line

File: Exiled.Events/Patches/Events/Scp173/AddingObserver.cs [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+                new(OpCodes.Callvirt, Method(typeof(HashSet<ReferenceHub>), nameof(HashSet<ReferenceHub>.Remove))),
+                new(OpCodes.Pop),
+
+                // return 0;
+                new(OpCodes.Ldc_I4_0),
+                new(OpCodes.Ret),
+            });
+
+            for (int z = 0; z < newInstructions.Count; z++)
+            {
```
Comment (NaoUnderscore) [inline Exiled.Events/Patches/Events/Scp173/AddingObserver.cs:82]: No braces

File: Exiled.Events/Handlers/Scp173.cs [hunk 14]
```diff
@@ -37,6 +37,16 @@ public static class Scp173
         /// </summary>
         public static Event<UsingBreakneckSpeedsEventArgs> UsingBreakneckSpeeds { get; set; } = new();
 
+        /// <summary>
+        /// Invoked before a user saw scp 173.
```
Comment (NaoUnderscore) [inline Exiled.Events/Handlers/Scp173.cs:41]:
```suggestion
        /// Invoked before a player starts looking at SCP-173.
```

File: Exiled.Events/Handlers/Scp173.cs [hunk 15]
```diff
@@ -37,6 +37,16 @@ public static class Scp173
...
+        /// </summary>
+        public static Event<AddingObserverEventArgs> AddingObserver { get; set; } = new();
+
+        /// <summary>
+        /// Invoked when a user no longer saw scp 173.
```
Comment (NaoUnderscore) [inline Exiled.Events/Handlers/Scp173.cs:46]:
```suggestion
        /// Invoked after a player stops looking at SCP-173.
```

File: Exiled.Events/Handlers/Scp173.cs [hunk 16]
```diff
@@ -60,5 +70,17 @@ public static class Scp173
         /// </summary>
         /// <param name="ev">The <see cref="UsingBreakneckSpeedsEventArgs" /> instance.</param>
         public static void OnUsingBreakneckSpeeds(UsingBreakneckSpeedsEventArgs ev) => UsingBreakneckSpeeds.InvokeSafely(ev);
+
+        /// <summary>
+        /// Called before a user is saw scp 173.
```
Comment (NaoUnderscore) [inline Exiled.Events/Handlers/Scp173.cs:75]:
```suggestion
        /// Called before a player starts looking at SCP-173.
```

File: Exiled.Events/Handlers/Scp173.cs [hunk 17]
```diff
@@ -60,5 +70,17 @@ public static class Scp173
...
+        /// </summary>
+        /// <param name="ev">The <see cref="AddingObserverEventArgs" /> instance.</param>
+        public static void OnAddingObserver(AddingObserverEventArgs ev) => AddingObserver.InvokeSafely(ev);
+
+        /// <summary>
+        /// Called when a user no longer saw scp 173.
```
Comment (NaoUnderscore) [inline Exiled.Events/Handlers/Scp173.cs:81]:
```suggestion
        /// Called after a player stops looking at SCP-173.
```

**Review by NaoUnderscore — CHANGES_REQUESTED** [review]
Keep the transpilers on the same patch as you're patching the same method, and it could lead to some issues, it's more maintainable in a single patch instead.

**Comment by skyyt15** [comment]
fuck

**Pull request closed without merging**